src/NuGet.Core/NuGet.Frameworks/CompatibilityProvider.cs

```diff
@@ -182,16 +182,39 @@ private bool IsCompatibleWithTarget(NuGetFramework target, NuGetFramework candid
 
         private static bool IsCompatibleWithTargetCore(NuGetFramework target, NuGetFramework candidate)
         {
-            bool result = NuGetFramework.FrameworkNameComparer.Equals(target, candidate)
-                    && IsVersionCompatible(target.Version, candidate.Version)
-                    && StringComparer.OrdinalIgnoreCase.Equals(target.Profile, candidate.Profile);
-
-            if (target.IsNet5Era && candidate.HasPlatform)
+            bool result = true;
+            bool isNet6Era = StringComparer.OrdinalIgnoreCase.Equals(FrameworkConstants.FrameworkIdentifiers.NetCoreApp, target.Framework) && target.Version.Major >= 6;
+            if (isNet6Era && target.HasPlatform && !NuGetFramework.FrameworkNameComparer.Equals(target, candidate))
             {
-                result = result
-                    && StringComparer.OrdinalIgnoreCase.Equals(target.Platform, candidate.Platform)
-                    && IsVersionCompatible(target.PlatformVersion, candidate.PlatformVersion);
+                if (candidate.Framework.StartsWith("xamarin.", StringComparison.OrdinalIgnoreCase))
+                {
+                    string suffix = candidate.Framework.ToUpperInvariant().Replace("XAMARIN.", "");
+                    var comp = StringComparer.OrdinalIgnoreCase;
+                    result = result &&
+                    ((comp.Equals(suffix, "mac") && comp.Equals(target.Platform, "macos"))
+                        || (comp.Equals(suffix, "ios") && comp.Equals(target.Platform, "ios"))
+                        || (comp.Equals(suffix, "tvos") && comp.Equals(target.Platform, "tvos"))
+                        || (comp.Equals(suffix, "ios") && comp.Equals(target.Platform, "maccatalyst")));
```

Comment:
I believe that checking compatibility is something on Restore's hot path. Is it worthwhile doing a benchmark to understand the impact of this change. In particular, Perhaps using a dictionary or hashset lookup rather than 4 comparisons might be faster?

Reply:
I've removed the allocations. One thing to note is that, in my experience, hash tables don't tend to be beneficial for lookups until you have 8+ or so items (depends on implementation). I don't know if that's the case for .NET but I don't really believe these checks would have a significant impact in performance.

If I were to use hash tables, I would have to go back to allocating (my grabbing the prefix). Is that enough, or would you still rather I come up with concrete benchmarks for this?

Reply:
You're right that hash table performance is complex, sometimes it's faster not to use them.

The thing is, I have no idea how many times in a real restore GetNearest is called, and therefore how impactful it is to overall restore time.

The cost of allocations for static data like this can be minimized by making it a static class field, so it's only allocated once per execution. Individual calls can still be zero allocation.

I haven't spent enough time thinking about your PR to have a good feeling about perf impact on an individual call level, especially since the changes are spread across two files. As previously mentioned, I don't know how often this methods are called.

@nkolev92 is our team's performance champ, so maybe he'll have different feedback, but maybe it's worthwhile at least doing a quick check with our `scripts/perftests` scripts to compare before and after. If we see a big perf hit, then we can spend more time thinking about perf improvements. If it has no impact, then we don't need to spend any time at all thinking about optimization. But right now I only have a gut feeling, with no evidence, that this has a real risk of being on a hot path.

Reply:
I wouldn't expect the perf scripts catch this change, but I don't have a good feel about how hot this codepath exactly is.
My guess is that there are 1000s of calls, so significant enough. 

Maybe the simple checkw ould be a microbenchmark in compat of things like net6 to netcoreapp/netstandard vs net6.0-platform to netcoreapp/netstandard vs net6.0-platform to netcoreapp/netstandard + xamarin, matching or not. 

If we start talking in `ms` or a huge difference in either case, it's worth double checking the performance in more detail on a restore level.